### ISSUE TITLE
Variable's address changed when passing them through DDP model

### ISSUE THREAD

**Issue by LukeForeverYoung** [pytorch/pytorch#82641]
### 🐛 Describe the bug

I passed a BatchEncoding, which is the output of a RobertaTokenizerFast tokenizer, through a distributed data parallel model. I find the variable address is changed and its ```.encodings``` attribute is lost.
The following is a minimize implementation:
``` python
import argparse
from torch import nn
import torch
from icecream import ic
import os
import torch.distributed as dist

from transformers.models.roberta.tokenization_roberta_fast import RobertaTokenizerFast

def setup_for_distributed(is_master):
        """
        This function disables printing when not in master process
        """
        import builtins as __builtin__

        builtin_print = __builtin__.print

        def print(*args, **kwargs):
            force = kwargs.pop("force", False)
            if is_master or force:
                builtin_print(*args, **kwargs)

        __builtin__.print = print
def init_distributed_mode(args):
    """Initialize distributed training, if appropriate"""
    if "RANK" in os.environ and "WORLD_SIZE" in os.environ:
        args.rank = int(os.environ["RANK"])
        args.world_size = int(os.environ["WORLD_SIZE"])
        args.gpu = int(os.environ["LOCAL_RANK"])
    elif "SLURM_PROCID" in os.environ:
        args.rank = int(os.environ["SLURM_PROCID"])
        args.gpu = args.rank % torch.cuda.device_count()
    else:
        print("Not using distributed mode")
        args.distributed = False
        return

    args.distributed = True

    torch.cuda.set_device(args.gpu)
    args.dist_backend = "nccl"
    print("| distributed init (rank {}): {}".format(args.rank, args.dist_url), flush=True)

    dist.init_process_group(
        backend=args.dist_backend, init_method=args.dist_url, world_size=args.world_size, rank=args.rank
    )
    dist.barrier()
    setup_for_distributed(args.rank == 0)

def get_args_parser():
    parser = argparse.ArgumentParser("Set transformer detector", add_help=False)
    # Distributed training parameters
    parser.add_argument("--world-size", default=1, type=int, help="number of distributed processes")
    parser.add_argument("--dist-url", default="env://", help="url used to set up distributed training")
    return parser



class M(nn.Module):
    def __init__(self) -> None:
        super().__init__()
        self.tokenizer = RobertaTokenizerFast.from_pretrained('ckpt/roberta-base')
        self.linear = nn.Linear(1,10)
    def forward(self,text=None,tokenized=None,xe=None):
        if tokenized:
            #tokenized._encodings = xe
            data = self.linear(tokenized.input_ids.float().unsqueeze(-1))
            return tokenized, data
        tokenized = self.tokenizer(text, padding="longest", return_tensors="pt").to('cuda')
        data = self.linear(tokenized.input_ids.float().unsqueeze(-1))
        return tokenized,data
        
if __name__=='__main__':
    args = get_args_parser().parse_args()

    init_distributed_mode(args)
    model = M()
    model.cuda()
    model = torch.nn.parallel.DistributedDataParallel(model, device_ids=[args.gpu])
    tokenizer = RobertaTokenizerFast.from_pretrained('ckpt/roberta-base')
    
    text = ['I dont know','I know']

    tokenized = tokenizer(text, padding="longest", return_tensors="pt").to('cuda')
    ic(tokenized._encodings)
    ic(id(tokenized))
    tokenized,_ = model(tokenized = tokenized)
    ic(tokenized._encodings)
    ic(id(tokenized))
```
The stdout is:
```
ic| tokenized._encodings: [Encoding(num_tokens=5, attributes=[ids, type_ids, tokens, offsets, attention_mask, special_tokens_mask, overflowing]),
                           Encoding(num_tokens=5, attributes=[ids, type_ids, tokens, offsets, attention_mask, special_tokens_mask, overflowing])]
ic| id(tokenized): 140585220520352
ic| tokenized._encodings: None
ic| id(tokenized): 140585219892224
```
This bug appears when I use pytorch 1.11.0 with cuda 11.3. But when I change to pytorch 1.10.1 with cuda 10.2, the bug disappears. The output is as follow:
```
ic| tokenized._encodings: [Encoding(num_tokens=5, attributes=[ids, type_ids, tokens, offsets, attention_mask, special_tokens_mask, overflowing]),
                           Encoding(num_tokens=5, attributes=[ids, type_ids, tokens, offsets, attention_mask, special_tokens_mask, overflowing])]
ic| id(tokenized): 140490778204240
ic| tokenized._encodings: [Encoding(num_tokens=5, attributes=[ids, type_ids, tokens, offsets, attention_mask, special_tokens_mask, overflowing]),
                           Encoding(num_tokens=5, attributes=[ids, type_ids, tokens, offsets, attention_mask, special_tokens_mask, overflowing])]
ic| id(tokenized): 140490778204240
```

## Env that has this bug
PyTorch version: 1.11.0
Is debug build: False
CUDA used to build PyTorch: 11.3
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.4 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: Could not collect
CMake version: version 3.14.0
Libc version: glibc-2.27

Python version: 3.8.13 (default, Mar 28 2022, 11:38:47)  [GCC 7.5.0] (64-bit runtime)
Python platform: Linux-4.19.91-009.ali4000.alios7.x86_64-x86_64-with-glibc2.17
Is CUDA available: True
CUDA runtime version: 11.0.194
GPU models and configuration: 
GPU 0: Tesla V100-SXM2-32GB
GPU 1: Tesla V100-SXM2-32GB

Nvidia driver version: 470.82.01
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.0.1
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.0.1
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.0.1
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.0.1
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.0.1
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.0.1
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.0.1
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.22.3
[pip3] torch==1.11.0
[pip3] torchaudio==0.11.0
[pip3] torchvision==0.12.0
[conda] blas                      1.0                         mkl  
[conda] cudatoolkit               11.3.1               h2bc3f7f_2  
[conda] ffmpeg                    4.3                  hf484d3e_0    pytorch
[conda] mkl                       2021.4.0           h06a4308_640  
[conda] mkl-service               2.4.0            py38h7f8727e_0  
[conda] mkl_fft                   1.3.1            py38hd3c417c_0  
[conda] mkl_random                1.2.2            py38h51133e4_0  
[conda] numpy                     1.22.3           py38he7a7128_0  
[conda] numpy-base                1.22.3           py38hf524024_0  
[conda] pytorch                   1.11.0          py3.8_cuda11.3_cudnn8.2.0_0    pytorch
[conda] pytorch-mutex             1.0                        cuda    pytorch
[conda] torchaudio                0.11.0               py38_cu113    pytorch
[conda] torchvision               0.12.0               py38_cu113    pytorch
## Env that do not has this bug
PyTorch version: 1.10.1
Is debug build: False
CUDA used to build PyTorch: 10.2
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.4 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: Could not collect
CMake version: version 3.14.0
Libc version: glibc-2.27

Python version: 3.8.13 (default, Mar 28 2022, 11:38:47)  [GCC 7.5.0] (64-bit runtime)
Python platform: Linux-4.19.91-009.ali4000.alios7.x86_64-x86_64-with-glibc2.17
Is CUDA available: True
CUDA runtime version: 11.0.194
GPU models and configuration: 
GPU 0: Tesla V100-SXM2-32GB
GPU 1: Tesla V100-SXM2-32GB

Nvidia driver version: 470.82.01
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.0.1
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.0.1
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.0.1
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.0.1
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.0.1
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.0.1
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.0.1
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.22.3
[pip3] torch==1.10.1
[pip3] torchaudio==0.10.1
[pip3] torchvision==0.11.2
[conda] blas                      1.0                         mkl  
[conda] cudatoolkit               10.2.89              hfd86e86_1  
[conda] ffmpeg                    4.3                  hf484d3e_0    pytorch
[conda] mkl                       2021.4.0           h06a4308_640  
[conda] mkl-service               2.4.0            py38h7f8727e_0  
[conda] mkl_fft                   1.3.1            py38hd3c417c_0  
[conda] mkl_random                1.2.2            py38h51133e4_0  
[conda] numpy                     1.22.3           py38he7a7128_0  
[conda] numpy-base                1.22.3           py38hf524024_0  
[conda] pytorch                   1.10.1          py3.8_cuda10.2_cudnn7.6.5_0    pytorch
[conda] pytorch-mutex             1.0                        cuda    pytorch
[conda] torchaudio                0.10.1               py38_cu102    pytorch
[conda] torchvision               0.11.2               py38_cu102    pytorch

cc @ezyang @gchanan @zou3519 @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang @kwen2501

**Comment by z-a-f** [comment]
A wild shot, can you try running it like so and see if there is an issue?:

```
from torch.nn.parallel._replicated_tensor_ddp_utils import _ddp_replicated_tensor
with _ddp_replicated_tensor(False):
    model = torch.nn.parallel.DistributedDataParallel(model, device_ids=[args.gpu])
    tokenizer = RobertaTokenizerFast.from_pretrained('ckpt/roberta-base')
    text = ['I dont know','I know']
    tokenized = tokenizer(text, padding="longest", return_tensors="pt").to('cuda')
```

**Comment by LukeForeverYoung** [comment]
> A wild shot, can you try running it like so and see if there is an issue?:
> 
> ```
> from torch.nn.parallel._replicated_tensor_ddp_utils import _ddp_replicated_tensor
> with _ddp_replicated_tensor(False):
>     model = torch.nn.parallel.DistributedDataParallel(model, device_ids=[args.gpu])
>     tokenizer = RobertaTokenizerFast.from_pretrained('ckpt/roberta-base')
>     text = ['I dont know','I know']
>     tokenized = tokenizer(text, padding="longest", return_tensors="pt").to('cuda')
> ```
I notice that the ```_ddp_replicated_tensor``` function has been available since pytorch 1.12. So I install torch1.12+cuda113 and the output is as follow:
```
ic| tokenized._encodings: [Encoding(num_tokens=5, attributes=[ids, type_ids, tokens, offsets, attention_mask, special_tokens_mask, overflowing]),
                           Encoding(num_tokens=5, attributes=[ids, type_ids, tokens, offsets, attention_mask, special_tokens_mask, overflowing])]
ic| id(tokenized): 140433083111936
ic| tokenized._encodings: None
ic| id(tokenized): 140432621859984
```
The modified code is:
```python
if __name__=='__main__':
    args = get_args_parser().parse_args()

    init_distributed_mode(args)
    with _ddp_replicated_tensor(False):
        model = M()
        model.cuda()
        model = torch.nn.parallel.DistributedDataParallel(model, device_ids=[args.gpu])
        tokenizer = RobertaTokenizerFast.from_pretrained('ckpt/roberta-base')
        
        text = ['I dont know','I know']

        tokenized = tokenizer(text, padding="longest", return_tensors="pt").to('cuda')
        ic(tokenized._encodings)
        ic(id(tokenized))
        tokenized,_ = model(tokenized = tokenized)
        ic(tokenized._encodings)
        ic(id(tokenized))
```
I also try the original script and this bug also appears when using pytorch 1.12.

**Comment by z-a-f** [comment]
> I also try the original script and this bug also appears when using pytorch 1.12.

To confirm:
- Works in 1.10
- Doesn't work in 1.11
- Doesn't work in 1.12 with `_ddp_replicated_tensor(False)`
- Doesn't work in 1.12 with `_ddp_replicated_tensor(True)`

**Comment by LukeForeverYoung** [comment]
> > I also try the original script and this bug also appears when using pytorch 1.12.
> 
> To confirm:
> 
> * Works in 1.10
> * Doesn't work in 1.11
> * Doesn't work in 1.12 with `_ddp_replicated_tensor(False)`
> * Doesn't work in 1.12 with `_ddp_replicated_tensor(True)`

Yes

**Comment by zhaojuanmao** [comment]
DDP does not change model states, it just synchronizes gradients. 

Would you please try to remove DDP wrapper and run a local training? and then see whether it still has the issue

**Comment by LukeForeverYoung** [comment]
> DDP does not change model states, it just synchronizes gradients.
> 
> Would you please try to remove DDP wrapper and run a local training? and then see whether it still has the issue

The issue disappears when the DDP wrapper is removed.

**Comment by zhaojuanmao** [comment]
I see, this may be related to _DDPSink pass through autograd node added in the DDP wrapper, this pass through autograd node may mutate the output in an unexpected way, we will investigate it further

**Comment by rohan-varma** [comment]
@zhaojuanmao given the way DDP is initialized `model = torch.nn.parallel.DistributedDataParallel(model, device_ids=[args.gpu])`, this shouldn't invoke _DDPSink right? We currently only pass it through when find_unused=True or static_graph=True?

**Comment by rohan-varma** [comment]
https://github.com/pytorch/pytorch/pull/68779/files seems _slightly_ suspicious, will take a look at it

**Comment by rohan-varma** [comment]
The change in https://github.com/pytorch/pytorch/pull/68779/files is indeed breaking this, I'll add a fix to revert the relevant snippet.
